pkg/data/leases.go

```diff
@@ -36,61 +37,116 @@ func (a *Lease) query(q *lease.Lease, keyName string, index string) (*lease.Leas
 	}
-	res, err = query(input, a.DynamoDB)
 
+	queryInput.SetLimit(*query.Limit)
+	if query.NextAccountID != nil && query.NextPrincipalID != nil {
+		// Should be more dynamic
+		queryInput.SetExclusiveStartKey(map[string]*dynamodb.AttributeValue{
+			"AccountId": &dynamodb.AttributeValue{
+				S: query.NextAccountID,
+			},
+			"PrincipalId": &dynamodb.AttributeValue{
+				S: query.NextPrincipalID,
+			},
+		})
+	}
+
+	res, err = a.DynamoDB.Query(queryInput)
 	if err != nil {
 		return nil, errors.NewInternalServer(
 			"failed to query leases",
 			err,
 		)
 	}
 
-	leases := &lease.Leases{}
-	err = dynamodbattribute.UnmarshalListOfMaps(res.Items, leases)
-	if err != nil {
-		return nil, errors.NewInternalServer("failed unmarshaling of leases", err)
-	}
-	return leases, nil
+	return &queryScanOutput{
+		items:            res.Items,
+		lastEvaluatedKey: res.LastEvaluatedKey,
+	}, nil
 }
 
-// scan for doing a scan against dynamodb
-func (a *Lease) scan(q *lease.Lease) (*lease.Leases, error) {
+// scanLeases for doing a scan against dynamodb
+func (a *Lease) scanLeases(query *lease.Lease) (*queryScanOutput, error) {
 	var expr expression.Expression
 	var err error
 	var res *dynamodb.ScanOutput
 
-	_, filters := getFiltersFromStruct(q, nil)
+	_, filters := getFiltersFromStruct(query, nil)
 	if filters != nil {
 		expr, err = expression.NewBuilder().WithFilter(*filters).Build()
 		if err != nil {
 			return nil, errors.NewInternalServer("unable to build query", err)
 		}
 	}
-	input := &dynamodb.ScanInput{
+
+	scanInput := &dynamodb.ScanInput{
 		TableName:                 aws.String(a.TableName),
 		ConsistentRead:            aws.Bool(a.ConsistentRead),
 		FilterExpression:          expr.Filter(),
 		ExpressionAttributeNames:  expr.Names(),
 		ExpressionAttributeValues: expr.Values(),
 	}
-	res, err = scan(input, a.DynamoDB)
-	if err != nil {
-		return nil, errors.NewInternalServer("error getting leases", err)
+
+	scanInput.SetLimit(*query.Limit)
+	if query.NextAccountID != nil {
+		// Should be more dynamic
+		scanInput.SetExclusiveStartKey(map[string]*dynamodb.AttributeValue{
+			"AccountId": &dynamodb.AttributeValue{
+				S: query.NextAccountID,
+			},
+			"PrincipalId": &dynamodb.AttributeValue{
+				S: query.NextPrincipalID,
+			},
+		})
 	}
 
-	leases := &lease.Leases{}
-	err = dynamodbattribute.UnmarshalListOfMaps(res.Items, leases)
+	res, err = a.DynamoDB.Scan(scanInput)
+
 	if err != nil {
-		return nil, errors.NewInternalServer("failed unmarshaling of leases", err)
+		return nil, errors.NewInternalServer("error getting leases", err)
 	}
-	return leases, err
+
+	return &queryScanOutput{
+		items:            res.Items,
+		lastEvaluatedKey: res.LastEvaluatedKey,
+	}, nil
 }
 
 // List Get a list of leases
 func (a *Lease) List(query *lease.Lease) (*lease.Leases, error) {
 
+	var outputs *queryScanOutput
+	var err error
+
+	if query.Limit == nil {
+		query.Limit = &a.Limit
+	}
+
 	if query.Status != nil {
```

Comment:
I think we can build out these scenarios based on the additional indexes for this table.  `PrincipalID`, `Id`?

Reply:
Fixed